### ISSUE TITLE
Inconsistency in the behavior of `create_mirror_view`

### ISSUE THREAD

**Issue by pzehner** [kokkos/kokkos#6842]
We observed a difference of behavior for `Kokkos::create_mirror_view` on the host memory space when the view to mirror is of constant items.

If `create_mirror_view` is called *without an execution space*, the mirrored view is duplicated, and its elements are not constant.
Otherwise, if it is called *with an execution space*, the mirrored view is the same as the input view, and its elements are constant.

Here is a MWE:

```cpp
// test.cpp

#include <cassert>
#include <Kokkos_Core.hpp>

void without_exec_space(Kokkos::View<const int*, Kokkos::HostSpace> a)
{
    // call without execution space
    auto b = Kokkos::create_mirror_view(a);

    static_assert(
            Kokkos::SpaceAccessibility<Kokkos::HostSpace, decltype(a)::memory_space>::accessible);

    // the mirrored view doesn't contain const ints anymore
    static_assert(std::is_same_v<decltype(b(0)), int&>);

    // views are different
    assert(b.data() != a.data());
}

void with_exec_space(Kokkos::View<const int*, Kokkos::HostSpace> a)
{
    // call with execution space
    auto b = Kokkos::create_mirror_view(Kokkos::DefaultHostExecutionSpace(), a);

    static_assert(std::is_same_v<
                  Kokkos::DefaultHostExecutionSpace::memory_space,
                  decltype(a)::memory_space>);
    static_assert(
            Kokkos::SpaceAccessibility<Kokkos::HostSpace, decltype(a)::memory_space>::accessible);

    // the mirrored view contains const ints
    static_assert(std::is_same_v<decltype(b(0)), const int&>);

    // views are the same
    assert(b.data() == a.data());
}

int main(int argc, char* argv[]) {
    Kokkos::initialize(argc, argv);
    {
        Kokkos::View<int*, Kokkos::HostSpace> a("A", 10);
        without_exec_space(a);
        with_exec_space(a);
    }
    Kokkos::finalize();
}
```

```cmake
# CMakeLists.txt

cmake_minimum_required(VERSION 3.18)

project(
    TestCreateMirrorView
    LANGUAGES CXX
)

find_package(Kokkos)

add_executable(
    test
    test.cpp
)

target_link_libraries(
    test
    PRIVATE
        Kokkos::kokkos
)
```

Any of the two behaviors may be valid, this is not the point of this issue, but only one should prevail.

We investigated the code and saw that the different behavior comes from the different implementation of the two signatures. For the first [call](https://github.com/kokkos/kokkos/blob/e2689abc05d2b96890148cc26b4a05c4180b8541/core/src/Kokkos_CopyViews.hpp#L3631), the code compares the input view data type to the `Kokkos::View<T, P...>::HostMirror` data type, where `HostMirror` is defined [here](https://github.com/kokkos/kokkos/blob/e2689abc05d2b96890148cc26b4a05c4180b8541/core/src/Kokkos_View.hpp#L651) and has the `traits::non_const_data_type` trait. Hence, the view is duplicated. For the second [call](https://github.com/kokkos/kokkos/blob/e2689abc05d2b96890148cc26b4a05c4180b8541/core/src/Kokkos_CopyViews.hpp#L3657), a simpler verification is done which only compares the execution space.

This is obviously a corner case, but a consistent behavior is important. Moreover, the [documentation](https://kokkos.org/kokkos-core-wiki/API/core/view/create_mirror.html) does not seem to indicate which behavior to expect. Namely, it does not document the non constant data type condition.

Consequently, what behavior should be expected? Is there some refactor to do?

I am interested to work on a corresponding PR, if any.

<details>
<summary>Extra details</summary>

1. Compilers: GCC 11.4.0
2. Kokkos release or commit used: develop branch (e2689abc05d2b96890148cc26b4a05c4180b8541)
3. Platform, architecture and backend: Linux x86_64, OpenMP and Serial backends
</details>

**Comment by masterleinad** [comment]
> If create_mirror_view is called without an execution space, the mirrored view is duplicated, and its elements are not constant.
Otherwise, if it is called with an execution space, the mirrored view is the same as the input view, and its elements are constant.

The use case for `create_mirror_view` is to use it with `deep_copy` to copy data from one memory space to another and it's OK for it to be a no-op in case the memory spaces coincide (or are accessible from the respective execution spaces).
You would use this function in case you want to fill the new View and then copy back or if the original View hasn't been filled with the appropriate data.

If the original `View` has a `const` data type, you can never modify its data through a View created from `create_mirror_view`. Thus, the only use case would be to copy its data to a different memory space. The problem is of course that
```
Kokkos::View<const double*, HostSpace> const_view;
Kokkos::create_mirror_view mirror_view = Kokkos::create_mirror_view(OtherMemorySpace{}, const_view);
Kokkos::deep_copy(mirror_view, const_view);
```
wouldn't compile if the `deep_copy` isn't a no-op. It seems it makes most sense to create a `non-const` View then.

**Comment by masterleinad** [comment]
We should discuss in the next meeting, though.

**Comment by dalg24** [comment]
Thanks for the good bug report.
I think we need to update the constraint in the unary `create_mirror_view` to compare the **non-const** data type to the host mirror data type.
We are definitely interested if you are willing to contribute a patch with a unit test that will prevent regressions.

**Comment by ajpowelsnl** [comment]
- Possibly related to:
  - #5924 
  - #5589

**Comment by crtrott** [comment]
hm interesting problem. If we allow `create_mirror_view` on const data to not return a new view, we may need to allow `deep_copy(const_a, const_a)`: i.e. allow deep_copy between two views of constant data if they are aliases.

**Comment by crtrott** [comment]
That said I am not sure if the current behavior is bad or not, we should think this through carefully.

**Comment by pzehner** [comment]
I add that, for GPU execution, the behavior of the two signatures is coherent.

<details>
<summary>See the modified MWE</summary>

```cpp
// test.cpp

#include <Kokkos_Core.hpp>

void without_exec_space(Kokkos::View<const int*> a)
{
    // call without execution space
    auto b = Kokkos::create_mirror_view(a);

    static_assert(
            !Kokkos::SpaceAccessibility<Kokkos::HostSpace, decltype(a)::memory_space>::accessible);

    // the mirrored view doesn't contain const ints anymore
    static_assert(std::is_same_v<decltype(b(0)), int&>);
}

void with_exec_space(Kokkos::View<const int*> a)
{
    // call with execution space
    auto b = Kokkos::create_mirror_view(Kokkos::DefaultHostExecutionSpace(), a);

    static_assert(!std::is_same_v<
                  Kokkos::DefaultHostExecutionSpace::memory_space,
                  decltype(a)::memory_space>);
    static_assert(
            !Kokkos::SpaceAccessibility<Kokkos::HostSpace, decltype(a)::memory_space>::accessible);

    // the mirrored view doesn't contain const ints either
    static_assert(std::is_same_v<decltype(b(0)), int&>);
}

int main(int argc, char* argv[]) {
    Kokkos::initialize(argc, argv);
    {
        Kokkos::View<int*> a("A", 10);
        without_exec_space(a);
        with_exec_space(a);
    }
    Kokkos::finalize();
}
```

</details>

**Comment by masterleinad** [comment]
Note that you are not actually using the execution spaces in your example but only their respective memory spaces to use an execution space instance, you would do
```
auto mirror_view = Kokkos::create_mirror_view(Kokkos::view_alloc{execution_space_instance, memory_space_instance}, view);
```
`Kokkos::create_mirror_view(Kokkos::DefaultHostExecutionSpace(), a);` does the same as `Kokkos::create_mirror_view(Kokkos::HostSpace(), a);` (since that was the original behavior before we introduced asynchronous overloads using execution space instances). Hence, you are really comparing specifying `Kokkos::HostSpace` explicitly vs. omitting it.

**Comment by pzehner** [comment]
Do you mean the signature that accept an execution space are asynchronous? This is not specified in the documentation, to my understanding of it.

**Comment by masterleinad** [comment]
> Do you mean the signature that accept an execution space are asynchronous? This is not specified in the documentation, to my understanding of it.

We are saying
> If a new [View()](https://kokkos.org/kokkos-core-wiki/API/core/view/view.html#_CPPv44Viewv) is created, the implicitly called constructor respects arg_prop and uses the same layout and padding as src.

**Comment by crtrott** [comment]
@pzehner apologies lol there may be some unfortunate legacy behavior we need to change to get stuff being fully consistent ...

**Comment by masterleinad** [comment]
@pzehner Do you have any use case for calling `create_mirror_view` with a `View` with `const` value type? 

The two alternatives we came up with yesterday were:
- disallow calling `create_mirror_view` in that case
- always create a new `View` (same behavior as calling `create_mirror`)

**Comment by pzehner** [comment]
Let me check with my colleagues. The only case I can think of is if you are inside a function where a View is passed with `const` values, by instance a function that computes a residual (i.e. cannot modify the source view).

**Comment by masterleinad** [comment]
> Let me check with my colleagues. The only case I can think of is if you are inside a function where a View is passed with `const` values, by instance a function that computes a residual (i.e. cannot modify the source view).

Right, so the question is why you would want to call `create_mirror_view` then.

**Comment by pzehner** [comment]
Mmh, imagine I create mirror views on the fly only when I need them (by instance in this residual function). But I would directly call `create_mirror_view_and_copy` then, I think.
